src/python/python-proto/python_proto/pipeline.py

```diff
@@ -1,111 +1,111 @@
+                proto_metadata.created_time
+            ).into_datetime(),
+            last_updated_time=Timestamp.from_proto(
+                proto_metadata.last_updated_time
+            ).into_datetime(),
         )
 
     def into_proto(self) -> _Metadata:
         proto_metadata = _Metadata()
         proto_metadata.trace_id.CopyFrom(Uuid.from_uuid(self.trace_id).into_proto())
         proto_metadata.tenant_id.CopyFrom(Uuid.from_uuid(self.tenant_id).into_proto())
+        proto_metadata.event_source_id.CopyFrom(
+            Uuid.from_uuid(self.event_source_id).into_proto()
+        )
+        proto_metadata.created_time.CopyFrom(
+            Timestamp.from_datetime(self.created_time).into_proto()
+        )
+        proto_metadata.last_updated_time.CopyFrom(
+            Timestamp.from_datetime(self.last_updated_time).into_proto()
+        )
         return proto_metadata
 
 
 @dataclasses.dataclass(frozen=True)
-class Envelope(SerDe):
+class Envelope(SerDeWithInner[_Envelope, I], Generic[I]):
     metadata: Metadata
-    inner_message: bytes
-    inner_type: str
+    inner_message: I
+    proto_cls: Type[_Envelope] = _Envelope
 
     @staticmethod
-    def deserialize(bytes_: bytes) -> Envelope:
+    def deserialize(bytes_: bytes, inner_cls: Type[I]) -> Envelope[I]:
         proto_envelope = _Envelope()
         proto_envelope.ParseFromString(bytes_)
-        return Envelope.from_proto(proto_envelope=proto_envelope)
+        return Envelope.from_proto(proto_envelope=proto_envelope, inner_cls=inner_cls)
 
     @staticmethod
-    def from_proto(proto_envelope: _Envelope) -> Envelope:
-        return Envelope._from_proto_parts(
-            proto_envelope.metadata,
-            proto_envelope.inner_message,
-            proto_envelope.inner_type,
+    def from_proto(proto_envelope: _Envelope, inner_cls: Type[I]) -> Envelope[I]:
+        inner_message = inner_cls.proto_cls()
+        proto_envelope.inner_message.Unpack(inner_message)
+        return Envelope(
+            metadata=Metadata.from_proto(proto_envelope.metadata),
+            inner_message=inner_cls.from_proto(inner_message),
         )
 
-    @staticmethod
-    def _from_proto_parts(
-        metadata: _Metadata,
-        inner_message: bytes,
-        inner_type: str,
-    ) -> Envelope:
-        return Envelope(Metadata.from_proto(metadata), inner_message, inner_type)
-
     def into_proto(self) -> _Envelope:
         proto_envelope = _Envelope()
         proto_envelope.metadata.CopyFrom(self.metadata.into_proto())
-        proto_envelope.inner_message = self.inner_message
-        proto_envelope.inner_type = self.inner_type
+        inner_message = _Any()
+        inner_message.Pack(
+            self.inner_message.into_proto(), type_url_prefix="".encode("utf-8")
+        )
+        proto_envelope.inner_message.CopyFrom(inner_message)
         return proto_envelope
+
+
+@dataclasses.dataclass(frozen=True)
+class RawLog(SerDe[_RawLog]):  # FIXME: test
```

Comment:
is this to be done asap or just soon

Reply:
This was actually done, I added strategies and tests for this type. Will remove comment.

Reply:
fixed